File: app/src/main/java/com/meiji/daily/util/DiffCallback.java
package com.meiji.daily.util;

import android.support.annotation.NonNull;
import android.support.v7.util.DiffUtil;

import me.drakeet.multitype.Items;
import me.drakeet.multitype.MultiTypeAdapter;

/**
 * Created by Meiji on 2017/6/7.
 */

public class DiffCallback extends DiffUtil.Callback {

    private final Items oldItems, newItems;

    private DiffCallback(Items oldItems, Items newItems) {
        this.oldItems = oldItems;
        this.newItems = newItems;
    }

    public static void create(@NonNull Items oldItems, @NonNull Items newItems, @NonNull MultiTypeAdapter adapter) {
        DiffCallback diffCallback = new DiffCallback(oldItems, newItems);
        DiffUtil.DiffResult result = DiffUtil.calculateDiff(diffCallback, true);
        result.dispatchUpdatesTo(adapter);
    }

    @Override
    public int getOldListSize() {
        return oldItems != null ? oldItems.size() : 0;
    }

    @Override
    public int getNewListSize() {
        return newItems != null ? newItems.size() : 0;
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return oldItems.get(oldItemPosition).equals(newItems.get(newItemPosition));
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return oldItems.get(oldItemPosition).hashCode() == newItems.get(newItemPosition).hashCode();
    }
}

File: app/src/main/java/com/meiji/daily/util/DiffCallback.kt
package com.meiji.daily.util

import android.support.v7.util.DiffUtil

import me.drakeet.multitype.Items
import me.drakeet.multitype.MultiTypeAdapter

/**
 * Created by Meiji on 2018/1/25.
 */

class DiffCallback private constructor(private val oldItems: Items?, private val newItems: Items?) : DiffUtil.Callback() {

    override fun getOldListSize(): Int {
        return oldItems?.size ?: 0
    }

    override fun getNewListSize(): Int {
        return newItems?.size ?: 0
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return oldItems!![oldItemPosition] == newItems!![newItemPosition]
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return oldItems!![oldItemPosition].hashCode() == newItems!![newItemPosition].hashCode()
    }

    companion object {

        fun create(oldItems: Items, newItems: Items, adapter: MultiTypeAdapter) {
            val diffCallback = DiffCallback(oldItems, newItems)
            val result = DiffUtil.calculateDiff(diffCallback, true)
            result.dispatchUpdatesTo(adapter)
        }
    }
}
